### PR TITLE
Skip Additional Rendering

### LINKED ISSUE #1919: ClientApp renders twice when following a react-router link
**Describe the bug**
The ClientApp is rendering multiple times when following a react router link, it appears to render with the current payload then follow the link and render again with the new payload. 

**To Reproduce**

1. Checkout [re-rendering](https://github.com/bbc/simorgh/pull/1920)
2. `npm run dev`
3. Visit http://localhost:7080/news/articles/c0g992jmmkko and click the link to the next optimo article `placed on the tomb of the unknown warrior`
4. Open console and see errors

**Expected behavior**
The ClientApp should only trigger one load of the App -> Article -> ArticleMain. Please add a test to `App.test.jsx` which checks we render the App once and once only. 

**Screenshots**

**Additional context**
This is causing issues where the ATI beacon is fired twice on a client page transition, the first beacon being the original page and the second beacon being the page being routed to. This makes me believe the ClientApp is re-rendering before following the react-router-link.

### PR DESCRIPTION
**Pull request by itimirichard**
Resolves #1919

**Overall change:** _Added a setTimeout to the first setState call inside the useEffect hook on App.jsx, this is to possibly skip the rerender caused by the first `setState` based on the user's internet speed. This `setState` primarily exists to set the `loading` value which just renders a white screen at the moment _

**Code changes:**

- _Wrap first setState in useEffect hook within an async setTimeout for 500ms._
- _clear setTimeout whenever route.getInitialData resolves._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [x] This PR requires manual testing

**Additional Context:**
Technically this is not a bug as the re-rendering is by design so that we can have a loading state (which will be handled by https://github.com/bbc/simorgh/issues/1248) . Until this is done, we can avoid the additional re-render (which just displays a blank white screen) for performance gains.

### CODE REVIEW

**Review by dr3 — CHANGES_REQUESTED** [review]
I think this may be blocked on https://github.com/bbc/simorgh/issues/1302 unfortunately

**Comment by itimirichard** [comment]
blocked on https://github.com/bbc/simorgh/issues/1302

File: src/app/containers/App/App.jsx [hunk 1]
```diff
@@ -59,20 +59,28 @@ export const App = ({ routes, location, initialData, history, headers }) => {
...
-        variant: nextVariant,
-        id: nextId,
-        isAmp: nextIsAmp,
-        pageType: route.pageType,
-        loading: true,
-        error: null,
-        errorCode: null,
+      let loaderTimeout;
+      const loaderPromise = new Promise(resolve => {
+        loaderTimeout = setTimeout(resolve, 500);
```
Comment (itimirichard) [inline src/app/containers/App/App.jsx:61]: I guess we can make the timeout value configurable

**Comment by itimirichard** [comment]
- For slow network speed 
Screenshot 2019-12-18 at 12 06 14 PM
- For fast network speed
Screenshot 2019-12-18 at 12 02 06 PM

**Comment by chris-hinds** [comment]
So this seems to work for the most part however I noticed that when the second render is skipped (due to the timeout) the users focus is still not reset when navigating to a new article i.e. the navigation happens and the article is still scrolled down half way where the user originally clicked.

I was under the impression that this was solved before this PR?

**Comment by itimirichard** [comment]
> So this seems to work for the most part however I noticed that when the second render is skipped (due to the timeout) the users focus is still not reset when navigating to a new article i.e. the navigation happens and the article is still scrolled down half way where the user originally clicked.
> 
> I was under the impression that this was solved before this PR?

Thanks @hindsc52 . This has been fixed by extending the parameters that trigger the focus i.e. `[state.loading, state.pageData]`. Previously, because the value of `loading` wasn't changing when the render is skipped, the `useEffect` that sets focus wasn't triggered.

File: src/app/containers/App/App.jsx [hunk 2]
```diff
@@ -56,28 +56,41 @@ export const App = ({ routes, location, initialData, bbcOrigin, history }) => {
...
+          variant: nextVariant,
+          id: nextId,
+          isAmp: nextIsAmp,
+          pageType: route.pageType,
           loading: false,
           pageData: path(['pageData'], data),
           status: path(['status'], data),
           error: path(['error'], data),
-        }));
+          errorCode: null,
```
Comment (itimirichard) [inline src/app/containers/App/App.jsx:92]: @hindsc52 I'd also like your thoughts on these changes. It looks fine to me, however I am bit skeptical about it. Thanks

Comment (chris-hinds) [reply src/app/containers/App/App.jsx:92]: How come this now includes service, variant, id etc instead of `...prevState`

Comment (itimirichard) [reply src/app/containers/App/App.jsx:92]: I realised for the case where the `setState` is skipped, `...prevState` doesn't then contain `nextService, nextVariant, nextId` etc and therefore the last`setState` is called with values from the previous article. So setting them explicitly ensures we maintain those values between the two `setState` calls even if one was skipped.

Comment (chris-hinds) [reply src/app/containers/App/App.jsx:92]: Thanks for the info @itimirichard

**Review by chris-hinds — APPROVED** [review]
Nice to see this feature is is now working, great work @itimirichard As this affects a core part of our application this will require manual testing before it gets merged.

**Review by jamesdonoh — COMMENTED** [review]
I'm coming to this late on so excuse me if I've missed this being discussed. Richard, thanks for figuring out a way of avoiding the flash of white screen. But if as you say this flash is by design, and since the issue you linked to (https://github.com/bbc/simorgh/issues/1248) is closed, unless there are imminent plans to resurrect this it did cross my mind whether a simpler solution would just be to remove the loading state entirely for now. Otherwise are we not just adding complexity working around a problem we have ourselves introduced?

**Comment by chris-hinds** [comment]
> I'm coming to this late on so excuse me if I've missed this being discussed. Richard, thanks for figuring out a way of avoiding the flash of white screen. But if as you say this flash is by design, and since the issue you linked to (#1248) is closed, unless there are imminent plans to resurrect this it did cross my mind whether a simpler solution would just be to remove the loading state entirely for now. Otherwise are we not just adding complexity working around a problem we have ourselves introduced?

I'd argue that we need to keep the loading state in place. We do not yet have a design for a loading page however the app still requires some sort of transition between rendering pages. If we were to remove the loading state those on slower networks wouldn't have any indication that a new page would even be loading.

At least with the blank page the user knows that something is happening (even though it may not be ideal)

**Review by jamesdonoh — APPROVED** [review]
Appreciate the explanation @hindsc52 - tbh I still don't entirely feel that a loading state is essential but as the fact is we already have one, and Richard has done well to mitigate the additional rendering issue, happy to merge this. Thanks for responding to my concerns.

**Comment by PriyaKR** [comment]
@itimirichard i can't see the focus being set on the h1 when the user navigates from one article to another article through an inline link.

**Comment by PriyaKR** [comment]
Looks good to me can be merged.Also the focus is being set on the H1 when the user navigates from one article to another article through an inline link.

**Comment by paruchurisilpa** [comment]
Looks good to me..

**Pull request merged**